README.md
# MMM-MoonPhase

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Displays the realtime moonphase and increments the shadow daily.

![Demo Image](https://github.com/NolanKingdon/MMM-MoonPhase/blob/master/images/2019-06-22-example-moons.png)

## Using the module

To install this module, navigate into `~/MagicMirror/modules` and type the following commands:

```sh
git clone https://github.com/NolanKingdon/MMM-MoonPhase
```

To use this module, add the following configuration block to the modules array in the `config/config.js` file:

```js
let config = {
	modules: [
		{
			module: "MMM-MoonPhase",
			position: "top_center",
			config: {
				updateInterval: 43200000,
				hemisphere: "N",
				resolution: "detailed",
				basicColor: "white",
				title: true,
				phase: true,
				x: 200,
				y: 200,
				alpha: 0.7
			}
		},
	]
}
```

## Configuration options

| Option           | Description                                       | Default              | Possible
|----------------- |---------------------------------------------------|----------------------|-----------
| `updateInterval` | *Optional* Miliseconds between updates            | `43200000` (12 hours)| any number
| `hemisphere`     | *Optional* Your location on the earth             | `"N"` (North)        | `"N"`/`"S"`
| `resolution`     | *Optional* Use detailed moon image or basic circle| `"detailed"`         | `"detailed"`/`"basic"`
| `basicColor`     | *Optional* Color of moon if using basic resolution| `"white"`            | any valid css color
| `title`          | *Optional* Toggle module title                    | `true`               | `true`/`false`
| `phase`          | *Optional* Toggle phase label                     | `true`               | `true`/`false`
| `x`              | *Optional* Width (recommended <300)               | `200`                | `>0`
| `y`              | *Optional* Height (recommended <300)              | `200`                | `>0`
| `alpha`          | *Optional* Visibility of dark side of moon        | `0.8` (Transparent)  | `>=0` to `<= 1`


CHANGELOG.md
# MMM-MoonPhase Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](https://semver.org/).

## [1.0.1] - Add translation

- Add fr translation to MMM-MoonPhase.js;
- Add sv translation to MMM-MoonPhase.js;
- CamelCase for variables;
- Use of var/let for variables;

## [1.0.0] - Unreleased

First public release


node_helper.js
/* MagicMirror²
 * Node Helper: MMM-MoonPhase
 *
 * By Nolan Kingdon
 * MIT Licensed.
 */

const NodeHelper = require("node_helper");

module.exports = NodeHelper.create({
	socketNotificationReceived: function(notification, payload) {
		switch(notification) {
			case 'CALCULATE_MOONDATA':
				const data = this.getMoonPhase(payload.x/2);
				this.sendSocketNotification('CURRENT_MOONDATA', data);
				break;
			default:
				console.error('Unexpected notification', notification, payload);
				break;
		}
	},
	getMoonPhase: function(radius){ // Gets the current Julian date
		/* NOTE - Math was heavily referenced from here:
		 *
		 * https://medium.com/@refik/a-journey-and-a-method-for-drawing-spheres-5b24246ca479
		 * https://github.com/refik/rotate-sphere
		 *
		 * A great writeup + the code tool he mentions in the article. These have been a great help
		 * to me.
		 */
		const currDate = new Date();
		let d = currDate.getDate();
		let m = currDate.getMonth()+1;
		let y = currDate.getFullYear();

		// Adjusting as per our formula found above
		if (m === 1 || m === 2){
		  y = y-1;
		  m = m + 12;
		}

		// Formula to determine number of new moons Julian dates
		const a   = y/100;
		const b   = a/4;
		const c   = 2-a+b;
		const e   = 365.25 * (y+4716);
		const f   = 30.6001 * (m+1);
		const jd  = c + d + e + f - 1524.5;
		const daysSinceNew  = jd - 2451549.5;
		const newMoons      = daysSinceNew / 29.53;
		const moonFraction  = "0." + newMoons.toString().split(".")[1];
		// Our final Digit - 29.53 days a moon cycle. 15 is full moon. 0/29.5 is new
		const dayOfCycle    = parseFloat(moonFraction * 29.53);
		const dayOfCycleMod = dayOfCycle%15; // Allows us to utilize the 180deg math below --> Use the same curve twice = less calc

		const points = [];
	  const rotateA = [0, 360*(dayOfCycleMod/29.5), 0]; // rotation angles
		const increaseBy = 2;
		
		for (i = 0; i < 180; i += increaseBy) {
			points.push(
				this.rotate(
					this.sphericalToPoint(90, i, radius),
					rotateA
				)
			);
		}

		// Returning our modded day AND our actual day
		return {
			points,
			jDate: [dayOfCycleMod, dayOfCycle]
		};
	}, 
	rotate: function(coor, rotateA) {
		return this.rotateForAxis(
			2, 
			this.rotateForAxis(
				1, 
				this.rotateForAxis(
					0, 
					coor,
					rotateA[0]
				),
				rotateA[1]
			), 
			rotateA[2]
		);
	},
	degreesToRadians: function(angle) {
		return angle * (Math.PI / 180);
	},
	// Convert spherical coordinate to x y z coordinate
	sphericalToPoint: function(ascension, declination, r) {
		ascension = this.degreesToRadians(ascension);
		declination = this.degreesToRadians(declination);
		return [
			Math.sin(ascension) * Math.sin(declination) * r,
			Math.cos(declination) * r,
			Math.cos(ascension) * Math.sin(declination) * r
		];
	},
	// Turn for x (0), y (1) or z (2) axis
	rotateForAxis: function(axis, coor, angle) {
		angle = this.degreesToRadians(angle);
		let coorStatic = coor.splice(axis, 1)[0];
		let c1 = coor[0];
		let c2 = coor[1];

		coor = [
			Math.cos(angle) * c1 - Math.sin(angle) * c2,
			Math.sin(angle) * c1 + Math.cos(angle) * c2
		];

		coor.splice(axis, 0, coorStatic);

		return coor;
	 },
	// Turn for all axis rotations
	rotate: function(coor, rotateA) {
		return this.rotateForAxis(
			2, 
			this.rotateForAxis(
				1, 
				this.rotateForAxis(
					0, 
					coor,
					rotateA[0]
				),
				rotateA[1]
			), 
			rotateA[2]
		);
	},
	// TODO -> Move all the math into here.
});


MMM-MoonPhase.js
/* MagicMirror²
 * Module: MMM-MoonPhase
 *
 * By Nolan Kingdon
 * MIT Licensed.
 */

Module.register("MMM-MoonPhase", {
	requiresVersion: "2.1.0", // Required version of MagicMirror
	defaults: { 
		updateInterval: 43200000, // Every Twelve hours
		hemisphere: "N", // N or S
		resolution: "detailed", // Detailed Or basic
		basicColor: "#FFF", // If basic moon (shape only) is used, this is it's color
		title: true, // Whether or not the Moon Phase Title is displayed
		phase: true, // Label for what moon phase it is
		age: false, // Display the age of the moon in days
		x: 200, // x dimension
		y: 200, // y dimension - I really recommend this staays the same as x, but whatever, go nuts
		alpha: 0.8 // Visibility of the moon behind the shadow - 1 is fully blacked out
	},
	start: function() {
		const self = this;
		this.moonData = {
			points: [],
			jDate: []
		}
		this.sendSocketNotification('CALCULATE_MOONDATA', this.config);

		// Schedule update timer.
		setInterval(function() {
			this.sendSocketNotification('CALCULATE_MOONDATA');
			// self.updateDom();
		}, this.config.updateInterval);
	},
	getStyles: function () {
		return [
			"MMM-MoonPhase.css",
		];
	},
	getTranslations: function() {
		/* If adding translations for the module,
		 * add the reference to the json file you
		 * created here!
		 */
		return {
			en: "translations/en.json",
			de: "translations/de.json",
			fr: "translations/fr.json",
			sv: "translations/sv.json",
			es: "translations/es.json",
			gr: "translations/gr.json",
			pl: "translations/pl.json",
			it: "translations/it.json",
			pt: "translations/pt-br.json",
			ru: "translations/ru.json",
			cs: "translations/cs.json"
		};
	},
	socketNotificationReceived: function (notification, payload) {
		if (notification === 'CURRENT_MOONDATA') {
			this.moonData = payload;
			this.updateDom();
		}
	},
	getDom: function() {
		const wrapper = document.createElement("div");
		const title = document.createElement("p");
		const jDate = this.moonData.jDate;
		wrapper.id = "moonphase-wrapper";
		title.id = "moonphase-title";

		if (this.config.title) {
			title.innerHTML = this.translate("TITLE");
		} else {
			title.style.display = "none";
		}

		const moonCanvas = document.createElement("canvas");
		moonCanvas.id = "moonphase-canvas";
		moonCanvas.height = this.config.y;
		moonCanvas.width = this.config.x;

		if (this.config.resolution === "detailed"){
			moonCanvas.style.background = "url('./MMM-MoonPhase/Phases/full.png?raw=true')";
			moonCanvas.style.backgroundSize = "cover";
		}

		const phase = document.createElement("p");
		phase.id = "moonphase-phase";

		if (jDate[1] < 1 || jDate[1] > 29){
			phase.innerHTML = this.translate("NEW");
		} else if (jDate[1] > 1 && jDate[1] < 7){
			phase.innerHTML = this.translate("WAX_CRESC");
		} else if (jDate[1] >= 7 && jDate[1] <= 8) {
			phase.innerHTML = this.translate("FIRST");
		} else if (jDate[1] > 8 && jDate[1] < 14){
			phase.innerHTML = this.translate("WAX_GIB");
		} else if (jDate[1] > 14 && jDate[1] < 16){
			phase.innerHTML = this.translate("FULL");
		} else if (jDate[1] > 16 && jDate[1] < 21){
			phase.innerHTML = this.translate("WAN_GIB");
		} else if (jDate[1] >= 22 && jDate[1] <= 23) {
			phase.innerHTML = this.translate("THIRD");
		} else if (jDate[1] > 23 && jDate[1] < 29){
			phase.innerHTML = this.translate("WAN_CRESC");
		}

		if (!this.config.phase){ 
			phase.style.display = "none";
		}

		const age = document.createElement("p");
		age.id  = "moonphase-age";

		if (!this.config.age){ 
			age.style.display = "none";
		}

		this.drawCanvas(age, moonCanvas);

		// Appending our elements to the DOM object
		wrapper.appendChild(title);
		wrapper.appendChild(moonCanvas);
		wrapper.appendChild(phase);
		wrapper.appendChild(age);
		return wrapper;
	},
	drawCanvas: function(age, canvas){
		const jDate = this.moonData.jDate;
		const ctx = canvas.getContext("2d");
		this.drawAxisCircles(jDate, ctx);

		// Transforming the moon image to align with the southern hemisphere
		if (this.config.hemisphere.toUpperCase() === "S"){
			canvas.style.transform = "rotate(180deg)";
		}

		age.innerHTML = Math.round(jDate[1]) + " " + this.translate("DAYS");
	},
	drawAxisCircles: function(jDate, ctx) {
		ctx.clearRect(0, 0, this.config.x, this.config.y);

		if (this.config.resolution === "basic"){
			ctx.beginPath();
			ctx.fillStyle = this.config.basicColor;
			ctx.arc(this.config.x/2, this.config.y/2, this.config.x/2, 1.5*Math.PI, 3.5 * Math.PI);
			ctx.fill();
			// Have to move back the cursor to not have interfering lines
			ctx.moveTo(this.config.x/2, this.config.y);
			ctx.closePath();
		}

		ctx.beginPath();

		// Draw curve
		for (i=0; i<this.moonData.points.length; i++) {
			ctx.fillStyle = "rgba(0,0,0, " + this.config.alpha + ")";
			this.drawPoint(
				this.moonData.points[i],
				ctx
			);
		}

		// Fill shadow
		if (jDate[1] < 15){ // We are in waxing phases
			// ClockWise
			ctx.arc(
				this.config.x/2,
				this.config.y/2,
				this.config.x/2,
				0.5*Math.PI,
				1.5*Math.PI
			);
			ctx.fill();
		} else if (jDate[1] >= 15){
			// Arc counter clckwise
			ctx.arc(
				this.config.x/2,
				this.config.y/2,
				this.config.x/2,
				0.5*Math.PI,
				1.5*Math.PI,
				true
			);
			ctx.fill();
		}
	},
	drawPoint: function(coor, ctx){
		// If the points z coordinate is less than
		// zero, it is out of view thus, grey.
		let color = "transparent";
		if (coor[2] >= 0) {
			color = "rgba(0,0,0, " + this.config.alpha + ")";
		}
		ctx.fillStyle = color;
		ctx.lineTo(this.config.x/2 + coor[0], this.config.y/2 - coor[1]);
		ctx.stroke();
	},
});
